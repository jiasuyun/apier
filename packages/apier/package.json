{
  "name": "@jiasuyun/apier",
  "version": "0.3.1",
  "description": "apier core",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "scripts": {
    "build": "tsc",
    "clean": "rimraf dist",
    "prepublish": "npm run -s clean && npm run -s build"
  },
  "files": [
    "dist"
  ],
  "publishConfig": {
    "access": "public"
  },
  "keywords": [],
  "author": "Sigoden Huang <sigoden@gmail.com> (https://sigoden.com/)",
  "homepage": "https://github.com/jiasuyun/apier",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/jiasuyun/apier"
  },
  "dependencies": {
    "@jiasuyun/apier-comment": "^0.3.1",
    "@jiasuyun/apier-parser-base": "^0.3.1",
    "@jiasuyun/apier-utils": "^0.3.0",
    "lodash": "^4.17.11"
  },
  "devDependencies": {
    "@jiasuyun/apier-parser-json5": "^0.3.1",
    "@jiasuyun/apier-test-utils": "^0.3.1",
    "@types/lodash": "^4.14.132",
    "rimraf": "^2.6.3",
    "typescript": "^3.4.5"
  }
}
