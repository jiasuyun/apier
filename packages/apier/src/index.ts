import {
  ApierRaws,
  ApierRaw,
  ApierRawObject,
  ApierRawReq,
  ApierRawRes,
  Method,
  Parser
} from "@jiasuyun/apier-parser-base";
import { ApierKind, kindOf } from "@jiasuyun/apier-utils";
import { ApierComment, CommentItem } from "@jiasuyun/apier-comment";
import lget from "lodash/get";

export abstract class ApierItem {
  public readonly comment: ApierComment;
  public readonly name: string;
  public readonly value: any;
  public readonly model: any;
  constructor(comment: ApierComment, name: string, value: any) {
    this.comment = comment.scope([name]);
    this.name = name;
    this.value = value;
  }
  public kind(): ApierKind {
    return kindOf(this.value);
  }
  public static createJSONKind(comment: ApierComment, name: string, value: any): ApierJSONKind {
    switch (kindOf(value)) {
      case ApierKind.INTEGER:
        return new ApierInteger(comment, name, value);
      case ApierKind.NUMBER:
        return new ApierNumber(comment, name, value);
      case ApierKind.STRING:
        return new ApierString(comment, name, value);
      case ApierKind.NULL:
        return new ApierNull(comment, name, value);
      case ApierKind.BOOLEAN:
        return new ApierBoolean(comment, name, value);
      case ApierKind.ARRAY:
        return new ApierArray(comment, name, value);
      case ApierKind.OBJECT:
        return new ApierObject(comment, name, value);
      default:
        throw new Error("unexpect undefined value");
    }
  }
}

export type ApierJSONKind = ApierArray | ApierObject | ApierParameter;
export type ApierParameter = ApierInteger | ApierString | ApierBoolean | ApierNull | ApierNumber;
export interface ApierReqModel {
  params?: ApierObject;
  query?: ApierObject;
  headers?: ApierObject;
  body?: ApierJSONKind;
}
export class ApierReq extends ApierItem {
  public readonly model: ApierReqModel;
  public readonly value: ApierRawReq;
  constructor(comment: ApierComment, name: string, value: ApierRawReq) {
    super(comment, name, value);
    this.model = Object.keys(value).reduce((model, key) => {
      model[key] = ApierItem.createJSONKind(this.comment, key, value[key]);
      return model;
    }, {});
  }
}
export interface ApierResModel {
  status: number;
  body?: ApierJSONKind;
}
export class ApierRes extends ApierItem {
  public readonly model: ApierResModel;
  public readonly value: ApierRawRes;
  constructor(comment: ApierComment, name: string, value: ApierRawRes) {
    super(comment, name, value);
    const { status, body } = value;
    this.model = { status };
    if (body) this.model.body = ApierItem.createJSONKind(this.comment, "body", body);
  }
}
export class ApierNumber extends ApierItem {
  public readonly model = null;
  public readonly value: number;
  constructor(comment: ApierComment, name: string, value: number) {
    super(comment, name, value);
  }
}

export class ApierInteger extends ApierItem {
  public readonly model = null;
  public readonly value: number;
  constructor(comment: ApierComment, name: string, value: number) {
    super(comment, name, value);
  }
}
export class ApierString extends ApierItem {
  public readonly model = null;
  public readonly value: string;
  constructor(comment: ApierComment, name: string, value: string) {
    super(comment, name, value);
  }
}
export class ApierBoolean extends ApierItem {
  public readonly model = null;
  public readonly value: boolean;
  constructor(comment: ApierComment, name: string, value: boolean) {
    super(comment, name, value);
  }
}
export class ApierNull extends ApierItem {
  public readonly model = null;
  public readonly value: null;
  constructor(comment: ApierComment, name: string, value: null) {
    super(comment, name, value);
  }
}
export type ApierArrayModel = ApierJSONKind[];
export class ApierArray extends ApierItem {
  public readonly model: ApierArrayModel;
  public readonly value: any[];
  constructor(comment: ApierComment, name: string, value: any[]) {
    super(comment, name, value);
    this.model = value.map((v, i) => ApierItem.createJSONKind(this.comment, String(i), v));
  }
}
export interface ApierObjectModel {
  [k: string]: ApierJSONKind;
}

export class ApierObject extends ApierItem {
  public readonly model: ApierObjectModel;
  public readonly value: ApierRawObject;
  constructor(comment: ApierComment, name: string, value: ApierRawObject) {
    super(comment, name, value);
    this.model = Object.keys(value).reduce((model, key) => {
      model[key] = ApierItem.createJSONKind(this.comment, key, value[key]);
      return model;
    }, {});
  }
}

export interface ApierModel {
  req?: ApierReq;
  res: ApierRes[];
}
export class Apier extends ApierItem {
  public readonly method: Method;
  public readonly url: string;
  public readonly value: ApierRaw;
  public readonly model: ApierModel;
  public readonly metadata: any;
  public readonly refs: Refs;
  constructor(comment: ApierComment, value: ApierRaw, metadata: any, refs: Refs) {
    super(comment, value.name, value);
    this.method = value.method;
    this.url = value.url;
    this.metadata = metadata;
    this.refs = refs;
    const model: any = {};
    if (value.req) model.req = new ApierReq(this.comment, "req", value.req);
    model.res = value.res.map((res, i) => new ApierRes(this.comment.scope(["res"]), "" + i, res));
    this.model = model;
  }
}

export interface ParseResult {
  apis: ApierRaws;
  comments: CommentItem[];
  apiers: Apier[];
  metadata: any;
}

export function parse(input: string, parser: Parser): ParseResult {
  const apiers = [];
  const { apis, comment, metadata } = parser.parse(input);
  const schemas = getRefs(apis, comment);
  for (const name in apis) {
    apiers.push(new Apier(comment, apis[name], metadata, schemas));
  }
  return { apis, apiers, comments: comment.comments, metadata };
}

export interface Refs {
  [k: string]: ApierJSONKind;
}

function getRefs(apis: ApierRaws, rootComment: ApierComment) {
  const refs: Refs = {};
  rootComment.comments.forEach(c => {
    const saveSchema = c.comment["saveSchema"];
    if (saveSchema) {
      const comment = rootComment.scope(c.paths.slice(0, -1));
      const tail = c.paths[c.paths.length - 1];
      refs[saveSchema] = ApierItem.createJSONKind(comment, tail, lget(apis, c.paths));
    }
  });
  return refs;
}
