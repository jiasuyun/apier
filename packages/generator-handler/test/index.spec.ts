import * as apier from "@jiasuyun/apier";
import Generator from "../src";
import { toYaml, factory } from "@jiasuyun/apier-test-utils-generator";

const gen = name => toYaml(factory((api: apier.Apier) => new Generator(api).value)(name));

test("generate", () => {
  expect(gen("general")).toMatchSnapshot();
});

test("expect(generate: real example)", () => {
  expect(gen("user")).toMatchSnapshot();
});
