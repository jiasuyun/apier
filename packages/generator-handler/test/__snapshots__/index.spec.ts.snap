// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect(generate: real example) 1`] = `
"- name: signup
  method: post
  url: /c/signup
  data:
    mock:
      id: '@string(\\"number\\", 5)'
      token: <jwt token>
  useMock: true
- name: loginPasswd
  method: post
  url: /c/login/passwd
  data:
    id: 32342
    token: <jwt token>
  useMock: false
- name: getGcode
  method: post
  url: /c/gcode
  data: <svg data>
  useMock: false
- name: loginSms
  method: post
  url: /c/login/sms
  data:
    id: 32342
    token: <jwt token>
  useMock: false
- name: getScode
  method: post
  url: /c/scode
  data:
    msg: OK
  useMock: false
- name: updatePasswd
  method: put
  url: /c/user/passwd
  data:
    msg: OK
  useMock: false
- name: listFavoriates
  method: get
  url: /c/favoriates
  data:
    mock|1-10:
      - type: 1
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          id: 33234
          content: how are you
      - type: 2
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          id: 32347
          thumbnail: <image url>
          name: 沙驰 系带舒适牛皮商务正装皮鞋男 21822007Z 黑色 40
          price: 232.34
      - type: 3
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          link: <href url>
      - type: 4
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          url: <image url>
      - type: 5
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          url: <image url>
      - type: 6
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          url: <image url>
      - type: 7
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          longitude: 22.5894947942
          latitude: 113.9739117838
          location: 广东省深圳市南山区桃源街道平山社区
      - type: 8
        id: 32134
        authorId: 3212
        authorName: 子鱼
        data:
          avatarUrl: <url>
  useMock: true
- name: checkProductFavoriate
  method: get
  url: '/c/favoriates/product/:id'
  data:
    isFavoriated: true
  useMock: false
- name: createFeedback
  method: post
  url: /c/feedbacks
  data:
    msg: OK
  useMock: false
- name: getStats
  method: get
  url: /c/stats
  data:
    numFollowings: 322
    numFavoriates: 100
    numQuns: 32
    numFriends: 320
  useMock: false
- name: getUserInfo
  method: get
  url: /c/user/info
  data:
    avatarUrl: <url>
    name: 子鱼
    myId: Brsadfng
    sex: 1
    regionId: 32423
    signature: <text>
  useMock: false
- name: updateUserInfo
  method: put
  url: /c/user/info
  data:
    msg: OK
  useMock: false
- name: getUserSettings
  method: get
  url: /c/user/settings
  data:
    notifyNewMessage: 1
    notifyPhoneCall: 1
    notifyShowDetail: 1
    notifyDynamic: 1
    notifyWithSound: 1
    notifyWithVibration: 1
    privacyNewFriend: 1
    privacyPhonenumSearchable: 1
    privacyIdSearchable: 1
    privacyRecommandContacts: 1
    commonBackground: <url>
  useMock: false
- name: updateUserSettings
  method: put
  url: /c/user/settings
  data:
    msg: OK
  useMock: false
"
`;

exports[`generate 1`] = `
"- name: getModel
  method: get
  url: '/model/:id'
  data:
    integer: 32
    number: 16.23
    bool: true
    'null': null
    array:
      - foo: 1
      - bar: 2
    object:
      password: a234324
  useMock: false
"
`;
